packages/inter-protocol/src/auction/auctionBook.js

```diff
@@ -0,0 +1,319 @@
+import '@agoric/zoe/exported.js';
+import '@agoric/zoe/src/contracts/exported.js';
+import '@agoric/governance/exported.js';
+
+import { M, makeScalarBigMapStore, provide } from '@agoric/vat-data';
+import { AmountMath, AmountShape } from '@agoric/ertp';
+import { Far } from '@endo/marshal';
+import { mustMatch } from '@agoric/store';
+import { observeNotifier } from '@agoric/notifier';
+
+import {
+  atomicRearrange,
+  ceilMultiplyBy,
+  floorDivideBy,
+  makeRatioFromAmounts,
+  multiplyRatios,
+  ratioGTE,
+} from '@agoric/zoe/src/contractSupport/index.js';
+import { TimeMath } from '@agoric/time';
+import { E } from '@endo/captp';
+import { makeTracer } from '@agoric/internal';
+
+import { makeDiscountBook, makePriceBook } from './discountBook.js';
+import {
+  AuctionState,
+  isDiscountedPriceHigher,
+  makeRatioPattern,
+} from './util.js';
+import { keyToTime } from './sortedOffers.js';
+
+const { Fail } = assert;
+
+/**
+ * @file The book represents the collateral-specific state of an ongoing
+ * auction. It holds the book, the lockedPrice, and the collateralSeat that has
+ * the allocation of assets for sale.
+ *
+ * The book contains orders for a particular collateral. It holds two kinds of
+ * orders: one has a price in terms of a Currency amount, the other is priced as
+ * a discount (or markup) from the most recent oracle price.
+ *
+ * Offers can be added in three ways. When the auction is not active, prices are
+ * automatically added to the appropriate collection. If a new offer is at or
+ * above the current price of an active auction, it will be settled immediately.
+ * If the offer is below the current price, it will be added, and settled when
+ * the price reaches that level.
+ */
+
+const trace = makeTracer('AucBook', false);
+
+const priceFrom = quote =>
+  makeRatioFromAmounts(
+    quote.quoteAmount.value[0].amountOut,
+    quote.quoteAmount.value[0].amountIn,
+  );
+
+/** @typedef {import('@agoric/vat-data').Baggage} Baggage */
+
+export const makeAuctionBook = async (
+  baggage,
+  zcf,
+  currencyBrand,
+  collateralBrand,
+  priceAuthority,
+) => {
+  const makeZeroRatio = () =>
+    makeRatioFromAmounts(
+      AmountMath.makeEmpty(currencyBrand),
+      AmountMath.make(collateralBrand, 1n),
+    );
+  const BidSpecShape = M.or(
+    {
+      want: AmountShape,
+      offerPrice: makeRatioPattern(currencyBrand, collateralBrand),
+    },
+    {
+      want: AmountShape,
+      offerDiscount: makeRatioPattern(currencyBrand, currencyBrand),
```

Comment:
> as you point out above this can be markup.

I first read that as "markup" as in HTML and was momentarily confused ;)

I can't think of one word for both either. But it seems like a concept that must have a word. Weird.